bindings_generator/src/special_methods.rs

```diff
@@ -289,6 +289,73 @@ impl std::ops::DerefMut for {name} {{
     Ok(())
 }
 
+pub fn generate_from_impl_for_upcast(
+    output: &mut impl Write,
+    api: &Api,
+    source: &GodotClass,
+    super_class_name: &str,
+) -> GeneratorResult {
+    if let Some(super_class) = api.find_class(super_class_name) {
+        if source.is_pointer_safe() {
+            writeln!(
+                output,
+                r#"
+// Upcast
+impl<'a> ::std::convert::From<&'a {source_name}> for {super_class_name} {{
+    fn from(obj: &'a {source_name}) -> Self {{
+        obj.to_{snake_case_super_class_name}()
+    }}
+}}
+"#,
+                source_name = source.name,
+                super_class_name = super_class.name,
+                snake_case_super_class_name = class_name_to_snake_case(super_class.name.as_str()),
+            )?;
+        } else {
+            writeln!(
+                output,
+                r#"
+// Ref upcast
+impl<'a> ::std::convert::From<&'a {source_name}> for {super_class_name} {{
+    fn from(obj: &'a {source_name}) -> Self {{
+        unsafe {{
+            obj.to_{snake_case_super_class_name}()
+        }}
+    }}
+}}
```

Comment:
Actually, I'm also not sure about this. This will allow one to create an alias without `unsafe`, which is decidedly not safe to do. I think reference upcasts should be restricted to reference-counted types.

Reply:
The appropriate traits should be `AsRef` and `Borrow` here.

Reply:
I think it may be necessary to define what these wrapper types conceptually represent. While all of their methods are `unsafe`, the distinction between shared and mutable references are seen in their APIs, so apparently we are still opting into *some* notion of safety here by creating one of the wrappers. We need to clarify what exactly are we promising to decide what is safe and what is not.

Reply:
In any case, as long as the types are as they are now (i.e. Not `Copy` or `Clone`), we likely don't want to alias them.

Reply:
A note, but this seems related to #175.

Looking  specifically at [upcasting](https://github.com/GodotNativeTools/godot-rust/blob/b2b9db0a376c62641e143518fe83ec81ee79c6f7/bindings_generator/src/special_methods.rs#L244-L247), note that none of this code actually relies on unsafe. No methods on Godot's end gets called if the pointer isn't ref-counted.

Namely, these are the two cases:
- parent is ref counted
  - call to ABI is guaranteed to be correct by the library -- if not, then we need to fix it
  - call to ABI needs a valid pointer. Here we have a further two cases
    - child is not ref-counted -> parent is valid (ref count) and child exists, so pointer must be valid (unless you can deallocate a child without deallocating a parent, but I think that that's a bug elsewhere)
    - child is ref-counted -> the pointer is valid
  - pointer aliasing is safe
- parent is not ref counted
  - no ABI call
  - pointer not necessarily valid, but since we don't use it, it's fine
  - pointer aliasing is considered safe
so, if my understanding of why we need `unsafe` is correct, upcasting is always safe.

Reply:
I agree that the upcasting of "owned" values is always safe. What I'm not sure about is the aliasing. It is true that the wrappers are in their implementation raw pointers, and bare raw pointers are safe to alias. However, they are conceptually not bare raw pointers. As mentioned above, the distinction between shared and mutable references is seen in the APIs. If they were conceptually meant to be bare pointers, then the distinction wouldn't have existed.

This is more related to #70 and #108, compared to the question you linked, imo.

I'm not entirely sure what the intentions were when the APIs were designed the way it currently is. Maybe @karroffel can shed some light on this?

Reply:
Oh, my bad, I thought that you were saying that it wasn't good to alias the pointer types because they were unsafe, not the mutability guarantees. I think aliasing is fine, though, if you don't change the mutability and ownership semantics (mutable reference to mutable reference only, reference to reference only), so not sure why you crossed out the initial comment -- `AsRef` and `AsMut` still work.

I agree that the `to_parent` methods are more of a `new_ref` method at the moment, so they don't really make sense. Would it be a good idea to replace them with an invocation of the Deref instead? Like so:

```rust
fn as_parent(&self) -> &Parent {
    &*self
}
fn as_parent_mut(&mut self) -> &mut Parent {
    &mut *self
}
```

and for the `cast` methods as well, if possible. `new_ref` is still available if they need to make a reference counted copy.

Reply:
> why you crossed out the initial comment

It's crossed out because the reason was formulated in a wrong way, not because the suggested traits were invalid.

> why the distinction wouldn't exist if they were just bare pointers

Pointer mutability doesn't mean anything. A `*const` is interchangeable with a `*mut`: you can simply cast between them and nothing will be broken. References, in contrast, come with strong guarantees, making it UB to cast a `&` to a `&mut`. Read dtolnay's *Accurate mental model for Rust's reference types* for more info: https://docs.rs/dtolnay/0.0.6/dtolnay/macro._02__reference_types.html

> replace them with an invocation of the Deref instead

I think it is better than the current API, but it probably should be in another PR. For this one it's probably better to just remove the aliasing impls.

Reply:
Well, if pointer mutability doesn't mean anything, then I can transitively say that references don't mean anything as well since references can be coerced to pointers. Something like this:
```rust
#[derive(Debug)]
struct Alpha(u64);

fn main() {
    let a: Alpha = Alpha(5);
    println!("{:?}", a);
    let b: &Alpha = &a;
    let c: *const Alpha = c;
    let c: *mut Alpha = c as *mut _;
    let c: &mut Alpha = unsafe { &mut *c };
    c.0 += 1;
    println!("{:?}, {:?}", a, e);
}
```
I can't directly cast between then, but I can convert between the two by relying on pointers. I did need to use `unsafe`, but (to me at least) that's the same as typing out `as *mut _` -- we're being trusted to not arbitrarily convert `*const` to `*mut` unless we know what we're doing. Both conversions need explicit mutability casts. So to me, pointers have the same meaning of "exclusivity", it's just on the developer to maintain those guarantees.

Also, before we end up too far off topic, I'm gathering that it would be good idea to split this into three tasks (read PRs):
1) Replacing `to_parent` and `cast` with alternatives more closely aligned with Rust semantics (once confirmed that this is indeed the case, and not something we're thinking too hard about).
2) AsRef/AsMut/Borrow impls for upcasting (this can be done as as extension of the `Deref` traits, so it shouldn't be necessary to wait)
3) implement examples for signal (from #163)

Reply:
> I can't directly cast between then, but I can convert between the two by relying on pointers. ... So to me, pointers have the same meaning of "exclusivity", it's just on the developer to maintain those guarantees.

No. What you're doing in that code example is UB: https://doc.rust-lang.org/nomicon/transmutes.html

> Transmuting an & to &mut is UB
>
> - Transmuting an & to &mut is always UB
> - No you can't do it
> - No you're not special

And no, just because you didn't call `transmute` doesn't mean that you are special.

`*mut` does not say anything about exclusivity, just like regular pointers in other languages with them. [It downright implements `Copy`](https://doc.rust-lang.org/std/primitive.pointer.html#impl-Copy). Converting a `*mut` to a `&mut` requires `unsafe` because it *is* unsafe: because it *adds* the guarantee of exclusivity. Which in turn also implies that the guarantee of exclusivity *wasn't* there.

- - -

> it would be good idea to split this into three tasks

I agree.